sdk/lib/rtl/heap.c

```diff
@@ -3488,77 +3538,112 @@ RtlpValidateHeap(PHEAP Heap,
+        /* Add up to the total amount of free entries */
+        FreeListEntriesCount++;
+    }
+
+    /* Check free list hints */
+    for (HintIndex = 0; HintIndex < Heap->DeCommitFreeBlockThreshold; HintIndex++)
+    {
+        if (Heap->FreeHints[HintIndex] != NULL)
         {
-            FreeEntry = CONTAINING_RECORD(NextEntry, HEAP_FREE_ENTRY, FreeList);
-            NextEntry = NextEntry->Flink;
+            PHEAP_FREE_ENTRY FreeEntry = CONTAINING_RECORD(Heap->FreeHints[HintIndex], HEAP_FREE_ENTRY, FreeList);
 
-            /* If there is an in-use entry in a free list - that's quite a big problem */
-            if (FreeEntry->Flags & HEAP_ENTRY_BUSY)
+            if (!RtlTestBit(&Heap->FreeHintBitmap, HintIndex))
             {
-                DPRINT1("HEAP: %Ix-dedicated list free element %p is marked in-use\n", Size, FreeEntry);
-                return FALSE;
+                DPRINT1("Hint bitmap bit at %u is not set, but there is a hint entry.\n", HintIndex);
             }
 
-            /* Check sizes according to that specific list's size */
-            if ((Size == 0) && (FreeEntry->Size < HEAP_FREELISTS))
-            {
-                DPRINT1("HEAP: Non dedicated list free element %p has size %x which would fit a dedicated list\n", FreeEntry, FreeEntry->Size);
-                return FALSE;
-            }
-            else if (Size && (FreeEntry->Size != Size))
+            if (HintIndex == 0)
             {
-                DPRINT1("HEAP: %Ix-dedicated list free element %p has incorrect size %x\n", Size, FreeEntry, FreeEntry->Size);
-                return FALSE;
+                 if (FreeEntry->Size <= Heap->DeCommitFreeBlockThreshold)
+                 {
+                     DPRINT1("There is an entry %p of size %lu, smaller than the decommit threshold %lu in the non-dedicated free list hint.\n",
+                             FreeEntry, FreeEntry->Size, Heap->DeCommitFreeBlockThreshold);
+                 }
             }
-            else if ((Size == 0) && (FreeEntry->Size < PreviousSize))
+            else
             {
-                DPRINT1("HEAP: Non dedicated list free element %p is not put in order\n", FreeEntry);
-                return FALSE;
-            }
-
-            /* Remember previous size*/
-            PreviousSize = FreeEntry->Size;
+                if (HintIndex != FreeEntry->Size - 1)
+                {
+                    DPRINT1("There is an entry %p of size %lu at the position %u in the free entry hint array.\n",
+                             FreeEntry, FreeEntry->Size, HintIndex);
+                }
 
-            /* Add up to the total amount of free entries */
-            FreeListEntriesCount++;
+                if (FreeEntry->FreeList.Blink != &Heap->FreeLists)
+                {
+                    /* The entry right before the hint must be smaller. */
+                    PHEAP_FREE_ENTRY PreviousFreeEntry = CONTAINING_RECORD(FreeEntry->FreeList.Blink,
```

Comment:
I prefer to declare variables where they are used instead of having that huge block of useless craft that we have everywhere.
And who am I to contradict Google ?
https://google.github.io/styleguide/cppguide.html#Local_Variables
> Place a function's variables in the narrowest scope possible, and initialize variables in the declaration.

Reply:
Yeah, that's the style guide which make the software bloated...